applications/IgaApplication/custom_conditions/sbm_laplacian_condition.cpp

```diff
@@ -0,0 +1,563 @@
+
+    void SBMLaplacianCondition::CalculateAllDirichlet(
+        MatrixType& rLeftHandSideMatrix,
+        VectorType& rRightHandSideVector,
+        const ProcessInfo& rCurrentProcessInfo,
+        const bool CalculateStiffnessMatrixFlag,
+        const bool CalculateResidualVectorFlag
+    )
+    {
+        Condition candidateClosestSkinSegment1 = this->GetValue(NEIGHBOUR_CONDITIONS)[0] ;
+
+        KRATOS_TRY
+        const auto& r_geometry = this->GetGeometry();
+        const SizeType number_of_nodes = r_geometry.PointsNumber();
+        if (rRightHandSideVector.size() != number_of_nodes) {
+            rRightHandSideVector.resize(number_of_nodes, false);
+        }
+        if (rLeftHandSideMatrix.size1() != number_of_nodes || rLeftHandSideMatrix.size2() != number_of_nodes) {
+            rLeftHandSideMatrix.resize(number_of_nodes, number_of_nodes, false);
+        }
+
+        noalias(rRightHandSideVector) = ZeroVector(number_of_nodes);
+        noalias(rLeftHandSideMatrix) = ZeroMatrix(number_of_nodes, number_of_nodes);
+
+        double penalty = GetProperties()[PENALTY_FACTOR];
+
+        // Integration
+        const GeometryType::IntegrationPointsArrayType& integration_points = r_geometry.IntegrationPoints();
+        const GeometryType::ShapeFunctionsGradientsType& DN_De = r_geometry.ShapeFunctionsLocalGradients(r_geometry.GetDefaultIntegrationMethod());
+        
+        // Initialize DN_DX
+        const unsigned int dim = DN_De[0].size2();
+        Matrix DN_DX(number_of_nodes,dim);
+        
+        Vector meshSize_uv = this->GetValue(MARKER_MESHES);
+        double h = std::min(meshSize_uv[0], meshSize_uv[1]);
+        if (dim == 3) {h = std::min(h,  meshSize_uv[2]);}
+
+        // Compute basis function order (Note: it is not allow to use different orders in different directions)
+        if (dim == 3) {
+            mbasisFunctionsOrder = std::cbrt(DN_De[0].size1()) - 1;
+        } else {
+            mbasisFunctionsOrder = std::sqrt(DN_De[0].size1()) - 1;
+        }
+
+        // Modify the penalty factor: p^2 * penalty / h (NITSCHE APPROACH)
+        penalty = mbasisFunctionsOrder * mbasisFunctionsOrder * penalty / h;
+
+        // Find the closest node in condition
+        auto GP_parameter_coord = r_geometry.Center();
+        int closestNodeId;
+        if (dim > 2) {
+            double incumbent_dist = 1e16;
+            for (int i = 0; i < dim; i++) {
+                if (norm_2(candidateClosestSkinSegment1.GetGeometry()[i]-GP_parameter_coord) < incumbent_dist) {
+                    incumbent_dist = norm_2(candidateClosestSkinSegment1.GetGeometry()[i]-GP_parameter_coord);
+                    closestNodeId = i;
+                }
+            }
+        }else{
+            closestNodeId = 0;
+        }
+        Vector projection(3);
+        projection[0] = candidateClosestSkinSegment1.GetGeometry()[closestNodeId].X() ;

```

Comment:
if you use `candidateClosestSkinSegment1.GetGeometry().Coordinates()` you call it only once